presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -143,111 +127,71 @@ private boolean advanceToNextRowGroup()
-        path.remove(ARRAY_TYPE_NAME);
-
-        if (elementOffsets.isEmpty()) {
-            for (int i = 0; i < batchSize; i++) {
-                elementOffsets.add(0);
-            }
-            return RunLengthEncodedBlock.create(elementType, null, batchSize);
-        }
-
-        int[] offsets = new int[batchSize + 1];
-        for (int i = 1; i < offsets.length; i++) {
-            offsets[i] = offsets[i - 1] + elementOffsets.getInt(i - 1);
-        }
-        return ArrayBlock.fromElementBlock(batchSize, new boolean[batchSize], offsets, block);
+        Field elementField = field.getChildren().get(0).get();
+        ColumnChunk columnChunk = readColumnChunk(elementField);
+        IntList offsets = new IntArrayList();
+        BooleanList valueIsNull = new BooleanArrayList();
+
+        ParquetListColumnReader.calculateCollectionOffsets(field, offsets, valueIsNull, columnChunk.getDefinitionLevels(), columnChunk.getRepetitionLevels());
+        Block arrayBlock = ArrayBlock.fromElementBlock(valueIsNull.size(), valueIsNull.toBooleanArray(), offsets.toIntArray(), columnChunk.getBlock());
+        return new ColumnChunk(arrayBlock, columnChunk.getDefinitionLevels(), columnChunk.getRepetitionLevels());
     }
 
-    public Block readMap(Type type, List<String> path)
+    private ColumnChunk readMap(GroupField field)
             throws IOException
     {
-        return readMap(type, path, new IntArrayList());
-    }
-
-    private Block readMap(Type type, List<String> path, IntList elementOffsets)
-            throws IOException
-    {
-        List<Type> parameters = type.getTypeParameters();
+        List<Type> parameters = field.getType().getTypeParameters();
         checkArgument(parameters.size() == 2, "Maps must have two type parameters, found %d", parameters.size());
         Block[] blocks = new Block[parameters.size()];
 
-        IntList keyOffsets = new IntArrayList();
-        IntList valueOffsets = new IntArrayList();
-        path.add(MAP_TYPE_NAME);
-        blocks[0] = readBlock(MAP_KEY_NAME, parameters.get(0), path, keyOffsets);
-        blocks[1] = readBlock(MAP_VALUE_NAME, parameters.get(1), path, valueOffsets);
-        path.remove(MAP_TYPE_NAME);
-
-        if (blocks[0].getPositionCount() == 0) {
-            for (int i = 0; i < batchSize; i++) {
-                elementOffsets.add(0);
-            }
-            return RunLengthEncodedBlock.create(parameters.get(0), null, batchSize);
-        }
-        int[] offsets = new int[batchSize + 1];
-        for (int i = 1; i < offsets.length; i++) {
-            int elementPositionCount = keyOffsets.getInt(i - 1);
-            elementOffsets.add(elementPositionCount * 2);
-            offsets[i] = offsets[i - 1] + elementPositionCount;
-        }
-        return ((MapType) type).createBlockFromKeyValue(new boolean[batchSize], offsets, blocks[0], blocks[1]);
+        ColumnChunk columnChunk = readColumnChunk(field.getChildren().get(0).get());
```

Comment:
This is because an array always has a single inner element and a map always has two inner elements (key and value). We do `isPresent` check for structs because some parameters can be missing due to schema evolution.